README.md
# defold-orthographic
Orthographic camera API for the [Defold game engine](https://www.defold.com). The API makes it super easy to convert screen to world coordinates, smoothly follow a game object and create a screen shake effect. This project is inspired by the camera component of the Phaser engine.

The project is shipped with an example that shows all the features of the orthographic camera. [Test the example app in your browser](http://britzl.github.io/Orthographic/index.html).

## Installation
You can use the orthograpic camera in your own project by adding this project as a [Defold library dependency](http://www.defold.com/manuals/libraries/). Open your game.project file and in the dependencies field under project add:

https://github.com/britzl/defold-orthographic/archive/master.zip

Or point to the ZIP file of a [specific release](https://github.com/britzl/defold-orthographic/releases).

## Basic usage
Add the ```camera.go``` to a collection. Depending on your use case you can either add the camera as a child of a game object to have the camera always follow that object or you could add the camera as a root game object and move or animate it manually using code or using the Orthographic Camera API (see below).

The camera will send view projection messages to the render script while it is enabled. Make sure your render script handles this message! See the section on render script integration below.

## Configuration
Select the script component attached to the ```camera.go``` to modify the properties. The camera has the following configurable properties:

#### near_z (number) and far_z (number)
This is the near and far z-values used in the projection matrix, ie the near and far clipping plane. Anything with a z-value inside this range will be drawn by the render script.

#### zoom (number)
This is the zoom level of the camera. Modify it by calling ```camera.zoom_to()```, ```go.set(camera, "zoom")``` or ```go.animate(camera, "zoom", ...)```. Read it using ```camera.get_zoom()``` or ```go.get(camera_id, "zoom")```.

#### projection (hash)
The camera can be configured to support different kinds of orthographic projections. The default projection (aptly named ```DEFAULT```) uses the same orthographic projection matrix as in the default render script (ie aspect ratio isn't maintained and content is stretched). Other projections are available out-of-the box:

* ```FIXED_AUTO``` - A fixed aspect ratio projection that automatically zooms in/out to fit the original viewport contents regardless of window size.
* ```FIXED_ZOOM``` - A fixed aspect ratio projection with zoom.

Note: For the above projections to work you need to pass the window dimensions from your render script to the camera. See [the section on render script integration](#render_script_integration).

Additional custom projections can be added, see ```camera.add_projector()``` below.

#### enabled (boolean)
This controls if the camera is enabled by default or not. Send ```enable``` and ```disable``` messages to the script or use ```go.set(id, "enable", true|false)``` to toggle this value.

#### follow (boolean)
This controls if the camera should follow a target or not. See ```camera.follow()``` for details.

#### follow_target (hash)
Id of the game object to follow. See ```camera.follow()``` for details.

#### follow_lerp (number)
Amount of lerp when following a target. See ```camera.follow()``` for details.

#### bounds_left (number), bounds_right (number), bounds_top (number), bounds_bottom (number)
The camera bounds. See ```camera.bounds()``` for details.

#### deadzone_left (number), deadzone_right (number), deadzone_top (number), deadzone_bottom (number)
The camera deadzone. See ```camera.deadzone()``` for details.


## Render script integration
While the camera is enabled it will send ```set_view_projection``` messages once per frame to the render script. The message is the same as that of the camera component, meaning that it contains ```id```, ```view``` and ```projection``` values. Make sure that these values are handled and used properly in the render script:

	function update(self)
		...
		render.set_view(self.view)
		render.set_projection(self.projection)
		-- draw using the view and projection
		...
	end

	function on_message(self, message_id, message, sender)
		if message_id == hash("set_view_projection") then
			self.camera_id = message.id
			self.view = message.view
			self.projection = message.projection
		end
	end

An alternative approach is to ignore the ```set_view_projection``` message and directly read the view and projection from the camera in the render script:

	local camera = require "orthographic.camera"

	function update(self)
		...
		local camera_id = id of your camera
		render.set_view(camera.get_view(camera_id))
		render.set_projection(camera.get_projection(camera_id))
		-- draw using the view and projection
		...
	end

NOTE: In order for this to work you need to make sure that the `Shared State` setting in the `Script` section of `game.project` is checked (defaults to checked)


### Feeding window size to the camera
It is recommended to send the window width and height from the render script to the camera module. This is required if any of the projectors provided in ```camera.lua``` is used. It also allows custom projectors to get the current window size by calling ```camera.get_window_size()```. Set the window size like this:

	local camera = require "orthographic.camera"

	function update(self)
		...
		local window_width = render.get_window_width()
		local window_height = render.get_window_height()
		camera.set_window_size(window_width, window_height)
		...
	end

NOTE: In order for this to work you need to make sure that the `Shared State` setting in the `Script` section of `game.project` is checked (defaults to checked)


### Example render script
The orthographic/render folder contains a render script that does the above mentioned integration of the Orthographic Camera API. Use it as it is or copy it into your project and make whatever modifications that you need.


## The Orthographic Camera API - functions
The API can be used in two ways:

1. Calling functions on the ```camera.lua``` module
2. Sending messages to the ```camera.script```

### camera.shake(camera_id, [intensity], [duration], [direction], [cb])
Shake the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```intensity``` (number) - Intensity of the shake, in percent of screen. Defaults to 0.05
* ```duration``` (number) - Duration of the shake, in seconds. Defaults to 0.5
* ```direction``` (hash) - Direction of the shake. Possible values: ```both```, ```horizontal```, ```vertical```. Defaults to ```both```.
* ```cb``` (function) - Function to call when the shake has finished. Optional.


### camera.stop_shaking(camera_id)
Stop shaking the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)


### camera.recoil(camera_id, offset, [duration])
Apply a recoil effect to the camera. The recoil will decay using linear interpolation.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```offset``` (vector3) - Offset to apply to the camera. Defaults to 0.05
* ```duration``` (number) - Duration of the recoil, in seconds. Defaults to 0.5


### camera.get_zoom(camera_id)
Get the current zoom level of the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)

**RETURN**
* ```zoom``` (number) The current zoom of the camera


### camera.zoom_to(camera_id, zoom)
Change the zoom level of the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```zoom``` (number) The new zoom level of the camera


### camera.follow(camera_id, target, [lerp])
Follow a game object.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```target``` (hash|url) - Game object to follow
* ```lerp``` (number) - Lerp from current position to target position with ```lerp``` as t. Optional.

### camera.unfollow(camera_id)
Stop following a game object.

**PARAMETERS**
* ```camera_id``` (hash|url)

### camera.deadzone(camera_id, left, top, right, bottom)
If following a game object this will add a deadzone around the camera position where the camera position will not update. If the target moves to the edge of the deadzone the camera will start to follow until the target returns within the bounds of the deadzone.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```left``` (number) - Number of pixels to the left of the camera
* ```top``` (number) - Number of pixels above the camera
* ```right``` (number) - Number of pixels to the right of the camera
* ```bottom``` (number) - Number of pixels below the camera

### camera.bounds(camera_id, left, top, right, bottom)
Limits the camera position to within the specified rectangle.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```left``` (number) - Left edge of the camera bounds
* ```top``` (number) - Top edge of camera bounds
* ```right``` (number) - Right edge of camera bounds
* ```bottom``` (number) - Bottom edge of camera bounds


### camera.screen_to_world(camera_id, screen)
Translate screen coordinates to world coordinates, based on the view and projection of the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)
* ```screen``` (vector3) Screen coordinates to convert

**RETURN**
* ```world_coords``` (vector3) World coordinates


### camera.screen_to_world_bounds(camera_id)
Translate screen boundaries (corners) to world coordinates, based on the view and projection of the camera.

**PARAMETERS**
* ```camera_id``` (hash|url)

**RETURN**
* ```bounds``` (vector4) Screen bounds (x = left, y = top, z = right, w = bottom)


### camera.world_to_screen(camera_id, world)
Translate world coordinates to screen coordinates, based on the view and projection of the camera. This is useful when manually culling game objects and you need to determine if a world coordinate will be visible or not.

**PARAMETER**
* ```camera_id``` (hash|url)
* ```world``` (vector3) World coordinates to convert

**RETURN**
* ```screen_coords``` (vector3) Screen coordinates


### camera.unproject(view, projection, screen)
Translate screen coordinates to world coordinates using the specified view and projection.

**PARAMETERS**
* ```view``` (matrix4)
* ```projection``` (matrix4)
* ```screen``` (vector3) Screen coordinates to convert

**RETURN**
* ```world_coords``` (vector3) Note: Same v3 object as passed in as argument


### camera.project(view, projection, world)
Translate world coordinates to screen coordinates using the specified view and projection.

**PARAMETERS**
* ```view``` (matrix4)
* ```projection``` (matrix4)
* ```world``` (vector3) World coordinates to convert

**RETURN**
* ```screen_coords``` (vector3) Note: Same v3 object as passed in as argument


### camera.add_projector(projector_id, projector_fn)
Add a custom projector that can be used by cameras in your project (see configuration above).

**PARAMETERS**
* ```projector_id``` (hash) - Id of the projector. Used as a value in the ```projection``` field of the camera script.
* ```projector_fn``` (function) - The function to call when a projection matrix is needed for the camera. The function will receive the id, near_z and far_z values of the camera.


### camera.use_projector(camera_id, projector_id)
Set a specific projector for a camera. This must be either one of the predefined projectors (see above) or a custom projector added using ```camera.add_projector()```.

**PARAMETERS**
* ```camera_id``` (hash) - Id of the camera to set projector for.
* ```projector_id``` (hash) - Id of the projector.


### camera.set_window_size(width, height)
Set the current window size so that it is available to projectors via ```camera.get_window_size()```. Set this via your render script.

**PARAMETERS**
* ```width``` (number) - Current window width.
* ```height``` (number) - Current window height.


### camera.get_window_size()
Get the current window size, as it was provided by ```camera.set_window_size()```. The default values will be the ones specified in game.project.

**RETURN**
* ```width``` (number) - Current window width.
* ```height``` (number) - Current window height.


### camera.get_display_size()
Get the display size, as specified in game.project.

**RETURN**
* ```width``` (number) - Display width.
* ```height``` (number) - Display height.

## The Orthographic Camera API - messages
Most of the functions of the API have message equivalents that can be sent to the camera component.

### shake
Message equivalent to ```camera.shake()```. Accepted message keys: ```intensity```, ```duration``` and ```direction```.

	msg.post("camera", "shake", { intensity = 0.05, duration = 2.5, direction = "both" })

### stop_shaking
Message equivalent to ```camera.stop_shaking()```.

	msg.post("camera", "stop_shaking")

### recoil
Message equivalent to ```camera.recoil()```. Accepted message keys: ```offset``` and ```duration```.

	msg.post("camera", "recoil", { offset = vmath.vector3(100, 100, 0), duration = 0.75 })

### shake_complete
Message sent back to the sender of a ```shake``` message when the shake has completed.

### follow
Message equivalent to ```camera.follow()```. Accepted message keys: ```target``` and ```lerp```.

	msg.post("camera", "follow", { target = hash("player"), lerp = 0.7 })

### unfollow
Message equivalent to ```camera.unfollow()```.

	msg.post("camera", "unfollow")

### deadzone
Message equivalent to ```camera.deadzone()```. Accepted message keys: ```left```, ```right```, ```bottom``` and ```top```.

	msg.post("camera", "deadzone", { left = 10, right = 200, bottom = 10, top = 100 })

### bounds
Message equivalent to ```camera.bounds()```. Accepted message keys: ```left```, ```right```, ```bottom``` and ```top```.

	msg.post("camera", "bounds", { left = 10, right = 200, bottom = 10, top = 100 })

### zoom_to
Message equivalent to ```camera.zoom_to()```. Accepted message keys: ```zoom```.

	msg.post("camera", "zoom_to", { zoom = 2.5 })

### enable
Enable the camera. While the camera is enabled it will update it's view and projection and send these to the render script.

	msg.post("camera", "enable")

### disable
Disable the camera.

	msg.post("camera", "disable")


CHANGELOG.md
## Orthographic Camera API 2.2.5 [britzl released 2018-04-16]
FIX: Assign default `display.width` and `display.height` (960x640) if none is provided from game.project

## Orthographic Camera API 2.2.4 [britzl released 2018-04-08]
FIX: An enabled camera is now immediately updated on creation. This will make sure that the render script gets the correct view and projection without any delay.

## Orthographic Camera API 2.2.3 [britzl released 2018-04-01]
FIX: Added missing message constant that made the camera.set_projection() function fail

## Orthographic Camera API 2.2.2 [britzl released 2018-03-31]
FIX: camera.set_zoom() crash

## Orthographic Camera API 2.2.1 [britzl released 2018-03-19]
FIX: Added check on script shared state in the provided render script. Also added a note on this in the readme

## Orthographic Camera API 2.2 [britzl released 2018-02-17]
NEW: Camera script properties to control bounds, deadzone, follow  
CHANGE: Camera script properties can now be manipulated using go.animate(), go.set() and go.get()

## Orthographic Camera API 2.1 [britzl released 2018-02-17]
NEW: camera.recoil()

## Orthographic Camera API 2.0 [britzl released 2018-02-17]
NEW: Camera zoom property. This makes it a whole lot easier to at run-time configure the zoom level of a camera  
NEW: camera.get_zoom() and camera.zoom_to()  
CHANGE: The available projections have been simplified. There's now only FIXED_AUTO and FIXED_ZOOM.  
  
Refer to the example project to see the new changes in action.

## Orthographic Camera API 1.5 [britzl released 2018-02-04]
NEW: Added camera.stop_shaking()

## Orthographic Camera API 1.4.2 [britzl released 2018-01-11]
FIX: Added argument asserts to all public facing functions on camera.lua

## Orthographic Camera API 1.4.1 [britzl released 2017-12-15]
FIX: Set an initial view and projection on the camera when it is initialized

## Orthographic Camera API 1.4 [britzl released 2017-12-15]
NEW: screen_to_world_bounds()

## Orthographic Camera API 1.3 [britzl released 2017-12-14]
NEW: camera.use_projector() to change projector at runtime  
NEW: camera.set_window_size() to feed current window size from render script to camera  
NEW: camera.get_window_size() to get current window size  
NEW: camera.get_display_size() to get display size from game.project  
NEW: camera.PROJECTORS.* constants for the provided projectors  
CHANGE: Moved projector functions from render script to camera.lua  


## Orthographic Camera API 1.2.2 [britzl released 2017-08-26]
* CHANGE: Moved the render script from the example project into the library folder.

## Orthographic Camera API 1.2.1 [britzl released 2017-08-03]
FIX: Bounds still didn't work as expected

## Orthographic Camera API 1.2 [britzl released 2017-08-01]
Delayed camera update to after all game objects have been update. This is to ensure that camera bounds are respected properly.

## Orthographic Camera API 1.1 [britzl released 2017-07-26]
* Fixed issues with the new bounds functionality  
* Added more projections (FIXED_NOZOOM, FIXED_ZOOM_2, FIXED_ZOOM_4, FIXED_ZOOM_6, FIXED_ZOOM_8 and FIXED_ZOOM_10)

## Orthographic Camera API 1.0 [britzl released 2017-07-24]
First official release of the Orthographic Camera API

## Orthographic Camera API 1.0 beta [britzl released 2017-06-30]
First public beta version of the API



orthographic/camera.lua
--- Camera module to use in combination with the camera.go or camera.script

local M = {}

M.SHAKE_BOTH = hash("both")
M.SHAKE_HORIZONTAL = hash("horizontal")
M.SHAKE_VERTICAL = hash("vertical")

M.PROJECTOR = {}
M.PROJECTOR.DEFAULT = hash("DEFAULT")
M.PROJECTOR.FIXED_AUTO = hash("FIXED_AUTO")
M.PROJECTOR.FIXED_ZOOM = hash("FIXED_ZOOM")

local DISPLAY_WIDTH = tonumber(sys.get_config("display.width")) or 960
local DISPLAY_HEIGHT = tonumber(sys.get_config("display.height")) or 640

local WINDOW_WIDTH = DISPLAY_WIDTH
local WINDOW_HEIGHT = DISPLAY_HEIGHT


-- center camera to middle of screen
local OFFSET = vmath.vector3(DISPLAY_WIDTH / 2, DISPLAY_HEIGHT / 2, 0)

local VECTOR3_ZERO = vmath.vector3(0)

local MATRIX4 = vmath.matrix4()

local v4_tmp = vmath.vector4()
local v3_tmp = vmath.vector3()

local cameras = {}

--- projection providers (projectors)
-- a mapping of id to function to calculate and return a projection matrix
local projectors = {}

-- the default projector from the default render script
-- will stretch content
projectors[M.PROJECTOR.DEFAULT] = function(camera_id, near_z, far_z, zoom)
	return vmath.matrix4_orthographic(0, DISPLAY_WIDTH, 0, DISPLAY_HEIGHT, near_z, far_z)
end

-- setup a fixed aspect ratio projection that zooms in/out to fit the original viewport contents
-- regardless of window size
projectors[M.PROJECTOR.FIXED_AUTO] = function(camera_id, near_z, far_z, zoom)
	local zoom_factor = math.min(WINDOW_WIDTH / DISPLAY_WIDTH, WINDOW_HEIGHT / DISPLAY_HEIGHT) * zoom
	local projected_width = WINDOW_WIDTH / zoom_factor
	local projected_height = WINDOW_HEIGHT / zoom_factor
	local xoffset = -(projected_width - DISPLAY_WIDTH) / 2
	local yoffset = -(projected_height - DISPLAY_HEIGHT) / 2
	return vmath.matrix4_orthographic(xoffset, xoffset + projected_width, yoffset, yoffset + projected_height, near_z, far_z)
end

-- setup a fixed aspect ratio projection with a fixed zoom
projectors[M.PROJECTOR.FIXED_ZOOM] = function(camera_id, near_z, far_z, zoom)
	local projected_width = WINDOW_WIDTH / zoom
	local projected_height = WINDOW_HEIGHT / zoom
	local xoffset = -(projected_width - DISPLAY_WIDTH) / 2
	local yoffset = -(projected_height - DISPLAY_HEIGHT) / 2
	return vmath.matrix4_orthographic(xoffset, xoffset + projected_width, yoffset, yoffset + projected_height, near_z, far_z)
end


--- Add a custom projector
-- @param projector_id Unique id of the projector (hash)
-- @param projector_fn The function to call when the projection matrix needs to be calculated
-- The function will receive near_z and far_z as arguments
function M.add_projector(projector_id, projector_fn)
	assert(projector_id, "You must provide a projector id")
	assert(projector_fn, "You must provide a projector function")
	projectors[projector_id] = projector_fn
end

--- Set the projector used by a camera
-- @param camera_id
-- @param projector_id The projector to use
function M.use_projector(camera_id, projector_id)
	assert(camera_id, "You must provide a camera id")
	assert(projector_id, "You must provide a projector id")
	local camera = cameras[camera_id]
	msg.post(camera.url, "use_projection", { projection = projector_id })
end


--- Set the window size
-- Call this from your render script to update the current window size
-- The width and height can later be retrieved through the M.get_window_size()
-- function. This is a convenience for use by custom projector functions
-- @param width Current window width
-- @param height Current window height
function M.set_window_size(width, height)
	assert(width, "You must provide window width")
	assert(height, "You must provide window height")
	WINDOW_WIDTH = width
	WINDOW_HEIGHT = height
end

--- Get the window size
-- @return width Current window width
-- @return height Current window height
function M.get_window_size()
	return WINDOW_WIDTH, WINDOW_HEIGHT
end

--- Get the display size (ie from game.project)
-- @return width Display width from game.project
-- @return height Display height from game.project
function M.get_display_size()
	return DISPLAY_WIDTH, DISPLAY_HEIGHT
end

local function calculate_projection(camera_id)
	local camera = cameras[camera_id]
	local projector_id = go.get(camera.url, "projection")
	local near_z = go.get(camera.url, "near_z")
	local far_z = go.get(camera.url, "far_z")
	local zoom = go.get(camera.url, "zoom")
	camera.zoom = zoom
	local projector_fn = projectors[projector_id] or projectors[hash("DEFAULT")]
	return projector_fn(camera_id, near_z, far_z, zoom)
end

local function calculate_view(camera_id, camera_world_pos, offset)
	local rot = go.get_world_rotation(camera_id)
	local pos = camera_world_pos - vmath.rotate(rot, OFFSET)
	if offset then
		pos = pos + offset
	end

	local look_at = pos + vmath.rotate(rot, vmath.vector3(0, 0, -1.0))
	local up = vmath.rotate(rot, vmath.vector3(0, 1.0, 0))
	local view = vmath.matrix4_look_at(pos, look_at, up)
	return view
end


--- Initialize a camera
-- Note: This is called automatically from the init() function of the camera.script
-- @param camera_id
-- @param camera_script_url
function M.init(camera_id, camera_script_url, settings)
	assert(camera_id, "You must provide a camera id")
	assert(camera_script_url, "You must provide a camera script url")
	cameras[camera_id] = settings
	cameras[camera_id].url = camera_script_url
	cameras[camera_id].view = calculate_view(camera_id, go.get_world_position(camera_id))	
	cameras[camera_id].projection = calculate_projection(camera_id)
end


--- Finalize a camera
-- Note: This is called automatically from the final() function of the camera.script
-- @param camera_id
function M.final(camera_id)
	assert(camera_id, "You must provide a camera id")
	cameras[camera_id] = nil
end


--- Update a camera
-- When calling this function a number of things happen:
-- * Follow target game object (if any)
-- * Limit camera to camera bounds (if any)
-- * Shake the camera (if enabled)
-- * Recalculate the view and projection matrix
--
-- Note: This is called automatically from the camera.script
-- @param camera_id
-- @param dt
function M.update(camera_id, dt)
	assert(camera_id, "You must provide a camera id")
	local camera = cameras[camera_id]
	if not camera then
		return
	end
	
	local camera_world_pos = go.get_world_position(camera_id)
	local camera_world_to_local_diff = camera_world_pos - go.get_position(camera_id)
	local follow_enabled = go.get(camera.url, "follow")
	if follow_enabled then
		local follow = go.get(camera.url, "follow_target")
		local target_pos = go.get_position(follow)
		local target_world_pos = go.get_world_position(follow)
		local new_pos
		local deadzone_top = go.get(camera.url, "deadzone_top")
		local deadzone_left = go.get(camera.url, "deadzone_left")
		local deadzone_right = go.get(camera.url, "deadzone_right")
		local deadzone_bottom = go.get(camera.url, "deadzone_bottom")
		if deadzone_top ~= 0 or deadzone_left ~= 0 or deadzone_right ~= 0 or deadzone_bottom ~= 0 then
			new_pos = vmath.vector3(camera_world_pos)
			local left_edge = camera_world_pos.x - deadzone_left
			local right_edge = camera_world_pos.x + deadzone_right
			local top_edge = camera_world_pos.y + deadzone_top
			local bottom_edge = camera_world_pos.y - deadzone_bottom
			if target_world_pos.x < left_edge then
				new_pos.x = new_pos.x - (left_edge - target_world_pos.x)
			elseif target_world_pos.x > right_edge then
				new_pos.x = new_pos.x + (target_world_pos.x - right_edge)
			end
			if target_world_pos.y > top_edge then
				new_pos.y = new_pos.y + (target_world_pos.y - top_edge)
			elseif target_world_pos.y < bottom_edge then
				new_pos.y = new_pos.y - (bottom_edge - target_world_pos.y)
			end
		else
			new_pos = target_world_pos
		end
		new_pos.z = camera_world_pos.z
		local follow_lerp = go.get(camera.url, "follow_lerp")
		camera_world_pos = vmath.lerp(follow_lerp, camera_world_pos, new_pos)
		camera_world_pos.z = new_pos.z
	end

	local bounds_top = go.get(camera.url, "bounds_top")
	local bounds_left = go.get(camera.url, "bounds_left")
	local bounds_bottom = go.get(camera.url, "bounds_bottom")
	local bounds_right = go.get(camera.url, "bounds_right")
	if bounds_top ~= 0 or bounds_left ~= 0 or bounds_bottom ~= 0 or bounds_right ~= 0 then
		local cp = M.world_to_screen(camera_id, vmath.vector3(camera_world_pos))
		local tr = M.world_to_screen(camera_id, vmath.vector3(bounds_right, bounds_top, 0)) - OFFSET
		local bl = M.world_to_screen(camera_id, vmath.vector3(bounds_left, bounds_bottom, 0)) + OFFSET
		
		cp.x = math.max(cp.x, bl.x)
		cp.x = math.min(cp.x, tr.x)
		cp.y = math.max(cp.y, bl.y)
		cp.y = math.min(cp.y, tr.y)
		
		camera_world_pos = M.screen_to_world(camera_id, cp)
	end

	go.set_position(camera_world_pos + camera_world_to_local_diff, camera_id)

	
	if camera.shake then
		camera.shake.duration = camera.shake.duration - dt
		if camera.shake.duration < 0 then
			camera.shake.cb()
			camera.shake = nil
		else
			if camera.shake.horizontal then
				camera.shake.offset.x = (DISPLAY_WIDTH * camera.shake.intensity) * (math.random() - 0.5)
			end
			if camera.shake.vertical then
				camera.shake.offset.y = (DISPLAY_WIDTH * camera.shake.intensity) * (math.random() - 0.5)
			end
		end
	end

	if camera.recoil then
		camera.recoil.time_left = camera.recoil.time_left - dt
		if camera.recoil.time_left < 0 then
			camera.recoil = nil
		else
			local t = camera.recoil.time_left / camera.recoil.duration
			camera.recoil.offset.x = vmath.lerp(t, 0, camera.recoil.offset.x)
			camera.recoil.offset.y = vmath.lerp(t, 0, camera.recoil.offset.y)
		end
	end

	local offset
	if camera.shake or camera.recoil then
		offset = VECTOR3_ZERO
		if camera.shake then
			offset = offset + camera.shake.offset
		end
		if camera.recoil then
			offset = offset + camera.recoil.offset
		end
	end
	camera.view = calculate_view(camera_id, camera_world_pos, offset)	
	camera.projection = calculate_projection(camera_id)
end


--- Follow a game object
-- @param camera_id
-- @param target The game object to follow
-- @param lerp Optional lerp to smoothly move the camera towards the target
function M.follow(camera_id, target, lerp)
	assert(camera_id, "You must provide a camera id")
	assert(target, "You must provide a target")
	msg.post(cameras[camera_id].url, "follow", { target = target, lerp = lerp })
end


--- Unfollow a game object
-- @param camera_id
function M.unfollow(camera_id)
	assert(camera_id, "You must provide a camera id")
	msg.post(cameras[camera_id].url, "unfollow")
end

--- Set the camera deadzone
-- @param camera_id
-- @param left Left edge of deadzone. Pass nil to remove deadzone.
-- @param top
-- @param right
-- @param bottom
function M.deadzone(camera_id, left, top, right, bottom)
	assert(camera_id, "You must provide a camera id")
	local camera = cameras[camera_id]
	if left and right and top and bottom then
		msg.post(camera.url, "deadzone", { left = left, top = top, right = right, bottom = bottom })
	else
		msg.post(camera.url, "deadzone")
	end
end


--- Set the camera bounds
-- @param camera_id
-- @param left Left edge of camera bounds. Pass nil to remove bounds.
-- @param top
-- @param right
-- @param bottom
function M.bounds(camera_id, left, top, right, bottom)
	assert(camera_id, "You must provide a camera id")
	local camera = cameras[camera_id]
	if left and top and right and bottom then
		msg.post(camera.url, "bounds", { left = left, top = top, right = right, bottom = bottom })
	else
		msg.post(camera.url, "bounds")
	end
end


--- Shake a camera
-- @param camera_id
-- @param intensity Intensity of the shake in percent of screen width. Optional, default: 0.05.
-- @param duration Duration of the shake. Optional, default: 0.5s.
-- @param direction both|horizontal|vertical. Optional, default: both
-- @param cb Function to call when shake has completed. Optional
function M.shake(camera_id, intensity, duration, direction, cb)
	assert(camera_id, "You must provide a camera id")
	cameras[camera_id].shake = {
		intensity = intensity or 0.05,
		duration = duration or 0.5,
		horizontal = direction ~= M.SHAKE_VERTICAL or false,
		vertical = direction ~= M.SHAKE_HORIZONTAL or false,
		offset = vmath.vector3(0),
		cb = cb,
	}
end


--- Stop shaking a camera
-- @param camera_id
function M.stop_shaking(camera_id)
	assert(camera_id, "You must provide a camera id")
	cameras[camera_id].shake = nil
end


--- Simulate a recoil effect
-- @param camera_id
-- @param offset Amount to offset the camera with
-- @param duration Duration of the recoil. Optional, default: 0.5s.
function M.recoil(camera_id, offset, duration)
	assert(camera_id, "You must provide a strength id")
	cameras[camera_id].recoil = {
		offset = offset,
		duration = duration or 0.5,
		time_left = duration or 0.5,
	}
end


--- Set the zoom level of a camera
-- @param camera_id
-- @param zoom The zoom level of the camera
function M.set_zoom(camera_id, zoom)
	assert(camera_id, "You must provide a camera id")
	assert(zoom, "You must provide a zoom level")
	msg.post(cameras[camera_id].url, "zoom_to", { zoom = zoom })
end


--- Get the zoom level of a camera
-- @param camera_id
-- @return Current zoom level of the camera
function M.get_zoom(camera_id)
	assert(camera_id, "You must provide a camera id")
	return cameras[camera_id].zoom
end


--- Get the projection matrix for a camera
-- @param camera_id
-- @return Projection matrix
function M.get_projection(camera_id)
	assert(camera_id, "You must provide a camera id")
	return cameras[camera_id].projection
end


--- Get the view matrix for a specific camera, based on the camera position
-- and rotation
-- @param camera_id
-- @return View matrix
function M.get_view(camera_id)
	assert(camera_id, "You must provide a camera id")
	return cameras[camera_id].view
end


--- Send the view and projection matrix for a camera to the render script
-- @param camera_id
function M.send_view_projection(camera_id)
	assert(camera_id, "You must provide a camera id")
	local view = cameras[camera_id].view or MATRIX4
	local projection = cameras[camera_id].projection or MATRIX4
	msg.post("@render:", "set_view_projection", { id = camera_id, view = view, projection = projection })
end


--- Convert screen coordinates to world coordinates based
-- on a specific camera's view and projection
-- @param camera_id
-- @param screen Screen coordinates as a vector3
-- @return World coordinates
-- http://webglfactory.blogspot.se/2011/05/how-to-convert-world-to-screen.html
function M.screen_to_world(camera_id, screen)
	assert(camera_id, "You must provide a camera id")
	assert(screen, "You must provide screen coordinates to convert")
	local view = cameras[camera_id].view or MATRIX4
	local projection = cameras[camera_id].projection or MATRIX4
	return M.unproject(view, projection, vmath.vector3(screen))
end


--- Convert world coordinates to screen coordinates based
-- on a specific camera's view and projection
-- @param camera_id
-- @param world World coordinates as a vector3
-- @return Screen coordinates
-- http://webglfactory.blogspot.se/2011/05/how-to-convert-world-to-screen.html
function M.world_to_screen(camera_id, world)
	assert(camera_id, "You must provide a camera id")
	assert(world, "You must provide world coordinates to convert")
	local view = cameras[camera_id].view or MATRIX4
	local projection = cameras[camera_id].projection or MATRIX4
	return M.project(view, projection, vmath.vector3(world))
end


--- Translate world coordinates to screen coordinates given a
-- view and projection matrix
-- @param view View matrix
-- @param projection Projection matrix
-- @param world World coordinates as a vector3
-- @return The mutated world coordinates (ie the same v3 object)
-- translated to screen coordinates
function M.project(view, projection, world)
	assert(view, "You must provide a view")
	assert(projection, "You must provide a projection")
	assert(world, "You must provide world coordinates to translate")
	v4_tmp.x, v4_tmp.y, v4_tmp.z, v4_tmp.w = world.x, world.y, world.z, 1
	local v4 = projection * view * v4_tmp
	world.x = ((v4.x + 1) / 2) * DISPLAY_WIDTH
	world.y = ((v4.y + 1) / 2) * DISPLAY_HEIGHT
	world.z = ((v4.z + 1) / 2)
	return world
end


local function unproject_xyz(inverse_view_projection, x, y, z)
	x = (2 * x / DISPLAY_WIDTH) - 1
	y = (2 * y / DISPLAY_HEIGHT) - 1
	z = (2 * z) - 1
	local inv = inverse_view_projection
	local x1 = x * inv.m00 + y * inv.m01 + z * inv.m02 + inv.m03
	local y1 = x * inv.m10 + y * inv.m11 + z * inv.m12 + inv.m13
	local z1 = x * inv.m20 + y * inv.m21 + z * inv.m22 + inv.m23
	return x1, y1, z1
end

--- Translate screen coordinates to world coordinates given a
-- view and projection matrix 
-- @param view View matrix
-- @param projection Projection matrix
-- @param screen Screen coordinates as a vector3
-- @return The mutated screen coordinates (ie the same v3 object)
-- translated to world coordinates
function M.unproject(view, projection, screen)
	assert(view, "You must provide a view")
	assert(projection, "You must provide a projection")
	assert(screen, "You must provide screen coordinates to translate")
	local inv = vmath.inv(projection * view)
	screen.x, screen.y, screen.z = unproject_xyz(inv, screen.x, screen.y, screen.z)
	return screen
end

--- Get the screen bounds as world coordinates, ie where in world space the
-- screen corners are
-- @param camera_id
-- @return bounds Vector4 where x is left, y is top, z is right and w is bottom
function M.screen_to_world_bounds(camera_id)
	assert(camera_id, "You must provide a camera id")
	local view = cameras[camera_id].view or MATRIX4
	local projection = cameras[camera_id].projection or MATRIX4
	local inv = vmath.inv(projection * view)
	local bl_x, bl_y = unproject_xyz(inv, 0, 0, 0)
	local br_x, br_y = unproject_xyz(inv, DISPLAY_WIDTH, 0, 0)
	local tl_x, tl_y = unproject_xyz(inv, 0, DISPLAY_HEIGHT, 0)
	return vmath.vector4(bl_x, tl_y, br_x, bl_y)
end

return M
